### PR TITLE
FdoSecrets: check and show PID and executable for existing secret service process

### PR DESCRIPTION
**Pull request by Aetf**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )

Check and show PID and executable for the existing secret service process. This should make it easier for users to debug if another service is running. See screenshots.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
70666433-9b756900-1c3c-11ea-8df2-8fe34b113079

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested manually. The warning message now shows the existing service PID.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by droidmonkey** [comment]
Very nice!

File: src/fdosecrets/objects/Service.cpp [hunk 1]
```diff
@@ -59,9 +59,11 @@ namespace FdoSecrets
...
     {
         if (!QDBusConnection::sessionBus().registerService(QStringLiteral(DBUS_SERVICE_SECRET))) {
-            qDebug() << "Another secret service is running";
-            emit error(tr("Failed to register DBus service at %1: another secret service is running.")
-                           .arg(QLatin1Literal(DBUS_SERVICE_SECRET)));
+            auto otherService = m_plugin->runningServiceInfo();
+            emit error(tr("Failed to register DBus service at %1: another secret service is running.<br/>")
+                           .arg(QLatin1Literal(DBUS_SERVICE_SECRET))
+                       + QStringLiteral("<i>%2</i><br/>").arg(otherService)
+                       + tr("Please stop/remove it before re-enabling the Secret Service Integration."));
```
Comment (phoerious) [inline src/fdosecrets/objects/Service.cpp]: You need to HTML-escape `otherService`.

File: src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp [hunk 2]
```diff
@@ -301,4 +312,53 @@ void SettingsWidgetFdoSecrets::saveSettings()
...
+}
+
+void SettingsWidgetFdoSecrets::checkDBusName()
+{
+    if (m_plugin->serviceInstance()) {
+        // only need checking if the service is not started or failed to start.
+        return;
+    }
+
+    qDebug() << "Checking DBus Name";
```
Comment (phoerious) [inline src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Do we need this?

File: src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp [hunk 3]
```diff
@@ -301,4 +312,53 @@ void SettingsWidgetFdoSecrets::saveSettings()
...
+    if (!reply.isValid()) {
+        showMessage(tr("Failed to connect to DBus. Please check your DBus setup."), MessageWidget::Error);
+        return;
+    }
+    if (reply.value()) {
+        auto otherService = m_plugin->runningServiceInfo();
+        showMessage(tr("<b>Warning:</b> Another secret service is running:<br/>")
+                        + QStringLiteral("<i>%1</i><br/>").arg(otherService)
+                        + tr("Please stop/remove it before re-enabling the Secret Service Integration."),
+                    MessageWidget::Warning);
```
Comment (phoerious) [inline src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Need to escape this as well. I am also not so happy with duplicating such a long error message in two places. Can we consolidate this somehow?

File: src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp (hunk 2 above)
Comment (Aetf) [reply src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Oops, this was only needed for debugging.

File: src/fdosecrets/objects/Service.cpp (hunk 1 above)
Comment (Aetf) [reply src/fdosecrets/objects/Service.cpp]: :ok_hand:

File: src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp (hunk 3 above)
Comment (Aetf) [reply src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Me either. I thought about the wording, but couldn't come up with something better. They are two different things in the end: one is the fact that the service failed to start, with a reason; this one is a warning. Any suggestions?

**Comment by Aetf** [comment]
What about this:

In settings

image

When error
image

**Comment by phoerious** [comment]
That could work.

File: src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp [hunk 4]
```diff
@@ -263,6 +267,15 @@ SettingsWidgetFdoSecrets::SettingsWidgetFdoSecrets(FdoSecretsPlugin* plugin, QWi
...
     connect(m_ui->enableFdoSecretService, &QCheckBox::toggled, m_ui->tabWidget, &QTabWidget::setEnabled);
+
+    // background checking
+    m_checkTimer.setInterval(2000);
+    connect(&m_checkTimer, &QTimer::timeout, this, &SettingsWidgetFdoSecrets::checkDBusName);
+    connect(m_plugin, &FdoSecretsPlugin::secretServiceStarted, &m_checkTimer, &QTimer::stop);
+    connect(m_plugin,
+            &FdoSecretsPlugin::secretServiceStopped,
+            &m_checkTimer,
+            static_cast<void (QTimer::*)()>(&QTimer::start));
```
Comment (phoerious) [inline src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Just noticed this weird cast here. What's that for?

Comment (Aetf) [reply src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: `QTimer::start` has multiple overloads. With the new connect syntax, you have to use `qOverload` (c++14) or `QOverload` (c++11) to select the appropriate one ([doc](https://doc.qt.io/qt-5/signalsandslots-syntaxes.html#selecting-overloaded-signals-and-slots)). But those are only available in Qt 5.7+.

Comment (phoerious) [reply src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Ah, right. This is extremely ugly and I would almost say use the old SIGNAL/SLOT syntax. What do you think @droidmonkey?

Comment (droidmonkey) [reply src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Yes in these cases I prefer the old style syntax. When implicit casts are not possible Qt will automatically find the correct connection.

**Comment by Aetf** [comment]
Use old style syntax to connect `QTimer::start`

**Pull request merged**